### PR TITLE
update: revamp integrate into app

### PR DESCRIPTION
**Pull request by katia-energizer**
* Created a separate article Architect your KMM app
* Created a tutorial on how to Make your Android application work on iOS

### CODE REVIEW

File: topics/integrate-in-existing-app.md [hunk 1]
```diff
@@ -1,242 +1,423 @@
...
-    sealed class Model {
-        object Form : Model()
-        object Loading : Model()
-        data class Error(val code: Int): Model()
-        object GoToNextScreen : Model()
-    }
-    interface View {
-        fun displayModel(model: Model)
-    }
+3. Ignore all warnings about platform-dependent code and click **Continue**.
```
Comment (KaterinaPetrova) [inline topics/integrate-in-existing-app.md:125]: Maybe it's worth explaining why we have these warnings? Like "IDE notifies us that some classes and methods will be unavailable in the shared module, but we will fix this later" or smth like this

**Review by KaterinaPetrova — CHANGES_REQUESTED** [review]
Great job 👏🏻👏🏻👏🏻
And I loved the new title!

File: topics/integrate-in-existing-app.md [hunk 2]
```diff
@@ -1,242 +1,423 @@
...
 
-## Decide what to share
+## Prepare an environment for development
 
-An application includes two layers:
+1. [Set up your environment for KMM development](setup.md) by installing the necessary tools on a macOS.
 
-* The **frontend** that defines the user interface and its behavior, including animations and transitions.
-* The **backend** that defines the business logic, data management, network stack, and other components that support the frontend.
+   >* For completing specific steps in this tutorial you need a Mac with macOS. These steps include writing iOS-specific code and running an iOS application.
```
Comment (koshachy) [inline topics/integrate-in-existing-app.md:16]: It seems that the info here is too important to be inside the note:
- The info that you should have Mac with macOS should be earlier (or not wrapped in the note).
- If Android Studio 4.2 is a requirement, it shouldn't be in a note. 

I'd prefer to see in that way:

## Prepare an environment ...

To start the project, you should have a Mac with macOS to work with Xcode...

1. Set up your env. This tutorial requires AS 4.2.
 or 

2. Update your Android studio up to 4.2.

File: topics/integrate-in-existing-app.md [hunk 3]
```diff
@@ -1,242 +1,423 @@
...
-or the [Model-View-Intent (_MVI_)](#mvi-for-declarative-ui-frameworks) pattern. These patterns:
+3. Click **Finish**.
 
-*   Make a clear distinction between the UI and presentation layers.
-*   Are completely decoupled from the UI platform.
-*   Are easily testable without a UI environment.
+The wizard creates the KMM shared module, updates configuration files, and creates files with classes that demonstrate multiplatform benefits.
+You can learn more about the [KMM project structure](discover-kmm-project.md).
+    
+### Depend on the shared module in your Android application
```
Comment (koshachy) [inline topics/integrate-in-existing-app.md:77]: The heading sounds weird to me.

It seems like you can unite this section with the previous one — **Create a shared module ...**. 
The meaning will remain the same.

File: topics/integrate-in-existing-app.md [hunk 4]
```diff
@@ -1,242 +1,423 @@
...
 
-#### MVP for legacy UI frameworks {initial-collapse-state="collapsed"}
+3. Apply the changes by synchronizing the project. 
 
-_Model-View-Presenter_ (_MVP_) forces you to create an API for both the Presenter that receives inputs and the View that displays outputs, allowing 
-you to test each independently.
+    ![Synchronize the Gradle files](gradle-sync.png)
 
-Here is an example of a simple MVP presenter:
+4. To check that the shared module is successfully connected to your application, dump the `greeting()` function result to the log
```
Comment (koshachy) [inline topics/integrate-in-existing-app.md:96]: If this step is necessary, we can create a separate section to check, that everything works ok.

File: topics/integrate-in-existing-app.md [hunk 5]
```diff
@@ -1,242 +1,423 @@
...
-        this.view = view.apply { displayModel(lastModel) }
-    }
-    fun detach() { this.view = null }
+4. Remove Android-specific code by replacing it with cross-platform Kotlin code or connecting to Android-specific APIs using [`expect` and `actual` declarations](connect-to-platform-specific-apis.md).
 
-    sealed class Intent {
-        data class Login(val username: String, val password: String)
-            : Intent()
-    }
+#### Replace Android-specific code with cross-platform code {initial-collapse-state="collapsed"}
```
Comment (koshachy) [inline topics/integrate-in-existing-app.md:131]: If these steps are required, maybe you should:
- Put a colon to the end of the previous sentence
- Explain the steps explicitly, like "do the following" or smth like that

File: topics/integrate-in-existing-app.md [hunk 6]
```diff
@@ -1,242 +1,423 @@
...
+    ```kotlin
+    package com.jetbrains.simplelogin.shared
+    
+    import platform.Foundation.NSUUID
+    actual fun randomUUID(): String = NSUUID().UUIDString()
+    ```     
+
+For Android and iOS, Kotlin will use different platform-specific implementations.
+
+#### Run your cross-platform application
```
Comment (koshachy) [inline topics/integrate-in-existing-app.md:216]: It's better to be the third-level (###) heading, I guess. Now it's almost invisible.
With instruction on how to run and what the result we got.

File: topics/integrate-in-existing-app.md [hunk 7]
```diff
@@ -1,242 +1,423 @@
...
+
+You can rename the `simpleLoginIOS` directory to the `iosApp` directory for consistency with other top-level directories of your cross-platform project.
+
+![Renamed iOS project directory in Android Studio](ios-directory-renamed-in-as.png){width=194}
+
+### Compile the shared module into a framework for the iOS project
+
+To use Kotlin code in your iOS project, compile shared code into a `.framework`.
+
+* In Android Studio, run the `packForXcode` Gradle task in one of the following ways:
```
Comment (koshachy) [inline topics/integrate-in-existing-app.md:263]: Maybe you should explain the most convenient (popular) way to run.
The second one you can wrap in the note, like 
> You can also run the following command in the Terminal ...

File: topics/integrate-in-existing-app.md [hunk 8]
```diff
@@ -1,242 +1,423 @@
...
+
+   * Double-click the `packForXcode` task in the list of Gradle tasks.
+   
+The generated framework is stored in `shared/build/xcode-frameworks/`.
+
+### Connect the framework to your iOS project
+
+Once you have the framework, you can connect it to your iOS project in different ways.
+
+This tutorial assumes that you store code for Android and iOS applications in a single repository (monorepo).
```
Comment (koshachy) [inline topics/integrate-in-existing-app.md:279]: I believe this info should be earlier.

And here you can repeat it in that way:  **Since the tutorial assumes ...**

**Review by koshachy — COMMENTED** [review]
I didn't pay special attention to the content (I think there's room for improvement) and tried to look through the structure to find some inconsistencies.

Please, take a look

**Review by katia-energizer — COMMENTED** [review]
Replied to feedback

File: topics/integrate-in-existing-app.md (hunk 1 above)
Comment (KaterinaPetrova) [reply topics/integrate-in-existing-app.md:126]: Ok, I agree about the explanation. My intention was not to explain the warning, but _why_ is we get some. But maybe it's too much and we have pretty comprehensive explanations about platform-specific apis below.

I would rather keep the screenshot, because as a user I'm used to using screenshots to check if I'm following the right directions, like a save points) :D

File: topics/integrate-in-existing-app.md [hunk 9]
```diff
@@ -1,242 +1,423 @@
...
+   >* For this tutorial, use Android Studio 4.2 or higher.
+   >
+   {type="note"}
 
-KMM presumes that the best user experience is when it is native to the platform itself, and therefore we don’t recommend sharing it in KMM. 
-However, you can share the code for the UI behavior that defines what happens with any user interaction and how the frontend communicates 
-with the backend with the help of compatible architectural patterns.
+2. In Android Studio, create a new project from the version control: `https://github.com/Kotlin/kmm-integration-sample`
+    
+    > You can create a similar application yourself using the **Login Activity** template for a new project with some modifications (TODO LINK).
```
[the remark this answers is not in the document]
Comment (katia-energizer) [reply topics/integrate-in-existing-app.md]: Removed.

File: topics/integrate-in-existing-app.md (hunk 1 above)
Comment (katia-energizer) [reply topics/integrate-in-existing-app.md:122]: I kept the screenshot but without explanation.

**Review by KaterinaPetrova — APPROVED** [review]

File: topics/integrate-in-existing-app.md [hunk 10]
```diff
@@ -1,242 +1,415 @@
...
-* The **backend** that defines the business logic, data management, network stack, and other components that support the frontend.
+   >You will need a Mac with macOS to complete certain steps in this tutorial, which include writing iOS-specific code and running an iOS application.  
+   >These steps cannot be performed on other operating systems, such as Microsoft Windows. This is due to an Apple requirement.
+   >
+   {type="note"}
 
-KMM presumes that the best user experience is when it is native to the platform itself, and therefore we don’t recommend sharing it in KMM. 
-However, you can share the code for the UI behavior that defines what happens with any user interaction and how the frontend communicates 
-with the backend with the help of compatible architectural patterns.
+2. In Android Studio, create a new project from version control: `https://github.com/Kotlin/kmm-integration-sample`.
```
Comment (koshachy) [inline topics/integrate-in-existing-app.md:21]: minor: This sentence looks strange on the staging site because of the line break.

File: topics/integrate-in-existing-app.md [hunk 11]
```diff
@@ -1,242 +1,415 @@
...
-| Layer | Recommendation on sharing|
-| ----- | -------------- |
-| Business logic | **Yes** |
-| Platform access | **Yes/no**. You’ll still need to [use platform-specific APIs](connect-to-platform-specific-apis.md), but you can share the behavior.|
-| Frontend behavior (reaction to inputs & communication with the backend) | **Yes/no**. Consider these [architectural patterns](#mvp-for-legacy-ui-frameworks).|
-| User interface (including animations & transitions) | **No**. It needs to be platform-specific.|
+## Make your code cross-platform
 
-## Integrate KMM into an existing application
+To make your application work on iOS, you'll first make your code cross-platform, and then you’ll reuse your cross-platform code in a new iOS application.
```
Comment (koshachy) [inline topics/integrate-in-existing-app.md:29]: minor: To **make** .., you'll first **make** ...
maybe rephrase?

File: topics/integrate-in-existing-app.md [hunk 12]
```diff
@@ -1,242 +1,415 @@
...
+    ```kotlin
+    dependencies {
+        implementation project(':shared')
     }
-}
-```
+    ```
 
-Note the `commandView` and `lastCommand` mechanism, which allows a view to detach and re-attach, for example for configuration changes on Android.
+3. Synchronize the project to apply the changes.
```
Comment (koshachy) [inline topics/integrate-in-existing-app.md:89]: minor: Here and below, many actions require pressing buttons. Some cases are described explicitly, and some are not.
I guess this is a matter of saving space, but it looks a little inconsistent.

File: topics/integrate-in-existing-app.md [hunk 13]
```diff
@@ -1,242 +1,415 @@
...
+
+   ![Warnings about platform-dependent code](warnings-android-specific-code.png){width=450}
+
+4. Remove Android-specific code by replacing it with cross-platform Kotlin code or connecting to Android-specific APIs using [`expect` and `actual` declarations](connect-to-platform-specific-apis.md). See the following sections for details.
+
+#### Replace Android-specific code with cross-platform code {initial-collapse-state="collapsed"}
+
+To make your code work well on both Android and iOS, replace all JVM dependencies with Kotlin dependencies wherever possible.
+
+1.  In the `login()` function of the `LoginDataSource` class, replace `IOException`, which is not available in Kotlin, with `RuntimeException`.
```
Comment (koshachy) [inline topics/integrate-in-existing-app.md:132]: minor: Some sentences (like this one) contain much inline code and can be rewritten to make them a little easier to read.
It's an idea of how to improve the user experience in the future.

File: topics/integrate-in-existing-app.md [hunk 14]
```diff
@@ -1,242 +1,415 @@
...
+val fakeUser = LoggedInUser(java.util.UUID.randomUUID().toString(), "Jane Doe")
 ```
 
-Note the `displayModel` and `lastModel` mechanism, which allows a view to detach and re-attach, for example for configuration changes on Android.
+Since the Kotlin standard library doesn't provide functionality for generating UUIDs, you still need to use platform-specific functionality for this case.
+
+Provide the `expect` declaration for the `randomUUID()` function in the shared code and its `actual` implementations for each platform – Android and iOS – in the corresponding source sets.
+You can learn more about [connecting to platform-specific APIs](connect-to-platform-specific-apis.md).
+
+1. Remove the `java.util.UUID` from the common code:
```
Comment (koshachy) [inline topics/integrate-in-existing-app.md:180]: minor: .. the `java.util.UUID` _what?_ ...

File: topics/integrate-in-existing-app.md [hunk 15]
```diff
@@ -1,242 +1,415 @@
...
+
+    ```kotlin
+   val fakeUser = LoggedInUser(randomUUID(), "Jane Doe") 
+   ```
+
+1. Create a `Utils.kt` file in the `shared/src/commonMain` and provide the `expect` declaration:
```
Comment (koshachy) [inline topics/integrate-in-existing-app.md:186]: minor: in the `share/src/commonMain` _what?_ (folder/directory)

File: topics/integrate-in-existing-app.md [hunk 16]
```diff
@@ -1,242 +1,415 @@
...
+
+You can rename the `simpleLoginIOS` directory to `iosApp` for consistency with other top-level directories of your cross-platform project.
+
+![Renamed iOS project directory in Android Studio](ios-directory-renamed-in-as.png){width=194}
+
+### Compile the shared module into a framework for the iOS project
+
+To use Kotlin code in your iOS project, compile shared code into a `.framework`.
+
+* In Android Studio, run the `packForXcode` Gradle task in the Terminal:
```
Comment (koshachy) [inline topics/integrate-in-existing-app.md:261]: minor: Do we really need a list with one item?

File: topics/integrate-in-existing-app.md [hunk 17]
```diff
@@ -1,242 +1,415 @@
...
+
+![iOS project in Android Studio](ios-project-in-as.png){width=194}
+
+You can rename the `simpleLoginIOS` directory to `iosApp` for consistency with other top-level directories of your cross-platform project.
+
+![Renamed iOS project directory in Android Studio](ios-directory-renamed-in-as.png){width=194}
+
+### Compile the shared module into a framework for the iOS project
+
+To use Kotlin code in your iOS project, compile shared code into a `.framework`.
```
Comment (koshachy) [inline topics/integrate-in-existing-app.md:259]: minor: 
> a `.framework`

I'm not quite sure about this article

File: topics/integrate-in-existing-app.md [hunk 18]
```diff
@@ -1,242 +1,415 @@
...
+   ![Synchronize the Gradle files](gradle-sync.png)
+
+You will see the new run configuration **simpleLoginIOS** for running your iOS application right from Android Studio.
+
+![iOS run configuration](ios-run-configuration-simplelogin.png){width=200}
+
+![iOS application password error](iphone-password-error.png){width=300}
 
-Once you're done, run your application on iOS to ensure that it works correctly.  
+![Android application password error](android-password-error.png){width=300}
```
Comment (koshachy) [inline topics/integrate-in-existing-app.md:400]: minor: Maybe combine three pictures into one (putting phones side by side) to save space?

**Review by koshachy — APPROVED** [review]
Overall is fine! Great job!
Nothing critical, but there is room for improvements in details (not only the ones that I mentioned). I suppose we can return for them after the publishing.

File: topics/integrate-in-existing-app.md (hunk 11 above)
Comment (katia-energizer) [reply topics/integrate-in-existing-app.md:29]: good point. Copyeditors missed. Updated.

File: topics/integrate-in-existing-app.md (hunk 12 above)
Comment (katia-energizer) [reply topics/integrate-in-existing-app.md:89]: I've added here. As to other places, I tried to balance a description and describe explicitly steps where the user may go a wrong way. But you're right it's not consistent. I don't see a big problem if a user follows steps and gets a result. It's a tutorial, not a procedure in docs. I would keep it flexible and listen to feedback.

File: topics/integrate-in-existing-app.md (hunk 13 above)
Comment (katia-energizer) [reply topics/integrate-in-existing-app.md:132]: If you have ideas, please share. This variant came from copyeditors. I had a different one but it's not easy to read due to formatting, I agree. Maybe we could format inline code differently.

File: topics/integrate-in-existing-app.md (hunk 14 above)
Comment (katia-energizer) [reply topics/integrate-in-existing-app.md:180]: thx. updated

File: topics/integrate-in-existing-app.md (hunk 15 above)
Comment (katia-energizer) [reply topics/integrate-in-existing-app.md:186]: Updated in all three places.

File: topics/integrate-in-existing-app.md (hunk 16 above)
Comment (katia-energizer) [reply topics/integrate-in-existing-app.md:261]: MS uses a single-bullet list for one-step procedures. I think it's a good way to highlight.

File: topics/integrate-in-existing-app.md (hunk 17 above)
Comment (katia-energizer) [reply topics/integrate-in-existing-app.md:259]: We discussed with copyeditors.

File: topics/integrate-in-existing-app.md (hunk 18 above)
Comment (katia-energizer) [reply topics/integrate-in-existing-app.md:400]: I thought about this. Maybe combine or create a table. Need to check what works better on different devices. But I'd do this consistently in other KMM docs as well. We'll discuss with Vika.

File: topics/integrate-in-existing-app.md (hunk 10 above)
Comment (katia-energizer) [reply topics/integrate-in-existing-app.md:21]: There is no a line break in the source. This is because the code is long and it is non-breakable. Nothing to do here 🤷‍♀️
